attestation-ca/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+            aaguids: self.aaguids,
+        }
+    }
+}
+
+impl TryFrom<SerialisableAttestationCa> for AttestationCa {
+    type Error = OpenSSLErrorStack;
+
+    fn try_from(data: SerialisableAttestationCa) -> Result<Self, Self::Error> {
+        Ok(AttestationCa {
+            ca: x509::X509::from_der(&data.ca.0)?,
+            aaguids: data.aaguids,
+        })
+    }
+}
+
+impl TryFrom<&[u8]> for AttestationCa {
+    type Error = OpenSSLErrorStack;
+
+    fn try_from(data: &[u8]) -> Result<Self, Self::Error> {
+        Ok(AttestationCa {
+            ca: x509::X509::from_pem(data)?,
+            aaguids: Default::default(),
+        })
+    }
+}
+
+impl AttestationCa {
+    /// Retrieve the Key Identifier for this Attestation Ca
+    pub fn get_kid(&self) -> Result<Vec<u8>, OpenSSLErrorStack> {
+        self.ca
+            .digest(hash::MessageDigest::sha256())
+            .map(|bytes| bytes.to_vec())
+    }
+
+    /// Update the set of aaguids this Attestation CA allows. If an empty btreeset is provided then
+    /// this Attestation CA allows all Aaguids.
+    pub fn set_aaguids(&mut self, aaguids: BTreeSet<Uuid>) {
+        self.aaguids = aaguids;
+    }
+
+    /// Update the set of aaguids this Attestation CA allows by adding this AAGUID to the allowed
+    /// set.
+    pub fn insert_aaguid(&mut self, aaguid: Uuid) {
+        self.aaguids.insert(aaguid);
+    }
+
+    /// Create a customised attestation CA from a DER public key.
+    pub fn new_from_der(data: &[u8]) -> Result<Self, OpenSSLErrorStack> {
+        Ok(AttestationCa {
+            ca: x509::X509::from_der(data)?,
+            aaguids: BTreeSet::default(),
+        })
+    }
+}
+
+/// A list of AttestationCas and associated options.
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct AttestationCaList {
+    /// The set of CA's that we trust in this Operation
+    pub cas: BTreeMap<Base64UrlSafeData, AttestationCa>,
+}
+
+impl TryFrom<AttestationCa> for AttestationCaList {
+    type Error = OpenSSLErrorStack;
+
+    fn try_from(att_ca: AttestationCa) -> Result<Self, Self::Error> {
+        let mut new = Self::default();
+        new.insert(att_ca)?;
+        Ok(new)
+    }
+}
+
+impl TryFrom<&[u8]> for AttestationCaList {
+    type Error = OpenSSLErrorStack;
+
+    fn try_from(data: &[u8]) -> Result<Self, Self::Error> {
+        let mut new = Self::default();
+        let att_ca = AttestationCa::try_from(data)?;
+        new.insert(att_ca)?;
+        Ok(new)
+    }
+}
+
+impl TryFrom<&[(&[u8], Uuid)]> for AttestationCaList {
+    type Error = OpenSSLErrorStack;
+
+    fn try_from(iter: &[(&[u8], Uuid)]) -> Result<Self, Self::Error> {
+        let mut cas = BTreeMap::default();
+
+        for (der, aaguid) in iter {
+            let ca = x509::X509::from_der(der)?;
+
+            let kid = ca
+                .digest(hash::MessageDigest::sha256())
+                // Is there a better way to do this with fromIterator?
+                .unwrap();
```

Comment:
Propagate error rather than unwrap, this looks like it returns `OpenSSLErrorStack` anyway